강산아/programmers/week27/메뉴리뉴얼.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Author    : Kang San Ah
+ * Date      : 2024.12.05(Thu)
+ * Runtime   : 1 sec
+ * Memory    : 256 MB
+ * Algorithm : dfs
+ */
+
+
+import java.util.*;
+
+public class 메뉴리뉴얼 {
+    static HashMap<String, Integer> hm = new HashMap<>();
+    static ArrayList<String> answerList = new ArrayList<>();
+
+    public String[] solution(String[] orders, int[] course) {
+        String[] answer = {};
+
+        for (int i : course) {
+            for (String order : orders) {
+                char[] chars = order.toCharArray();
+                Arrays.sort(chars);
+
+                dfs(0, "", chars, i);
+            }
+
+            if (!hm.isEmpty()) {
+                List<Integer> countList = new ArrayList<>(hm.values());
+                int max = Collections.max(countList);
+
+                if (max > 1) {
+                    for (String key : hm.keySet()) {
+                        if (hm.get(key) == max) {
+                            answerList.add(key);
+                        }
+                    }
+                }
+            }
+            hm.clear();
+        }
+
+        Collections.sort(answerList);
+        answer = new String[answerList.size()];
+
+        for (int i = 0; i < answer.length; i++) {
+            answer[i] = answerList.get(i);
+        }
+
+        return answer;
+    }
+
+    static void dfs(int depth, String course, char[] chars, int i) {
+        if (course.length() == i) {
+            if (hm.containsKey(course)) {
+                hm.put(course, hm.get(course) + 1);
+            }else{
+                hm.put(course, 1);
+            }
```

Comment:
```suggestion
            /*
                1. hm.put(course, hm.getOrDefault(course, 0) + 1);
                2. hm.compute(course, (key, val) -> (val == null) ? 1 : val + 1);
                3. hm.computeIfAbsent(course, key -> 0);
                    hm.put(course, hm.get(course) + 1);
           */
```

제가 선호하는 스타일의 코드인데 취향적인 부분이지만 추천하는 나름의 이유는 있습니다.

1. if - else 의 Depth 제거 => 복잡성 감소
2. 람다식 이용 가능 => Lazy Evaluation 을 생각 했을 때 이점을 볼 수 있는 코드가 있음

이러한 이유로 저는 Map 객체를 사용할 때 해당 메소드들을 애용 하는데, 저도 이전에는 똑같은 방식으로 containsKey 를 먼저 호출한 뒤 존재 여부를 가지고 로직을 작성했기 때문에 혹시 도움이 될까 추천 드립니다.